README


This archive contains the distribution Limper,
version 0.012:

  extremely lightweight but not very powerful web application framework

This software is copyright (c) 2014 by Ashley Willis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Limper

0.012     2014-10-27 03:11:07+00:00 UTC
- skip tests for MSWin32
- add CONTRIBUTING and EVEN MORE sections to Limper
- add LAYOUT and NAMESPACES sections to Limper::Extending
- add placeholders Limper::Plugin and Limper::Engine for those namespaces
- add Limper::Passphrase as example of what not to make a plugin
- add Limper::Differences which compares against Dancer
- changed 'options' to 'config', since 'options' is an HTTP method
- handle HTTP methods OPTIONS and PATCH
- add 'any' to create handlers for all methods
- allow chaining the method functions (they all return @_ now)
- changed 'note' to 'info'
- better headers function, and change $response->{headers} to a HASH
- request->{header} is now what request->{hheader} was – no more ARRAY form

0.011     2014-10-20 05:46:31+00:00 UTC
- much better DESCRIPTION, and other POD fixes and tweaks

0.010     2014-10-20 05:06:01+00:00 UTC
- delete $response->{body} if $head, so that response_handler works correctly for HEAD requests

0.009     2014-10-20 02:48:07+00:00 UTC
- changes to support Limper::PSGI (and possibly Limper::SCGI, etc):
  - set $request->{remote_host} to $conn->peerhost // "localhost"
  - new exported subs response and options
  - new hooks response_handler and request_handler
  - tweaks to support proper use of new hooks

0.008     2014-10-19 06:06:40+00:00 UTC
- properly handle HTTP/1.0 close/keep-alive
- add POD Limper::Extending

0.007     2014-10-17 04:41:56+00:00 UTC
- fix bug when calling headers in list mode and it's undefined

0.006     2014-10-16 02:03:34+00:00 UTC
Work towards HTTP/1.1 compliance and be usable in production.

- HTTP 1.1 improvements:
  - require Host: header and accept absolute URLs from HTTP/1.1 clients
  - support chunked encoding and timeout on reading content.
  - return an RFC 1123 Date: header with every response
  - minimal 100 Continue support
  - better close support

- seems like a good idea:
  - catch exceptions instead of having worker die

- new/changed features:
  - after hook support – runs after all other processing, just before response is sent.
  - rename 'logg' to 'note', add similar 'warning'

0 general code improvements:
  - remove prototypes
  - use strict in example.pl
  - remove passing http version to send_response
  - add PodCoverageTests to dist.ini
  - make $conn package global
  - tweak warning and bad_request
  - tweak timeout

0.005     2014-10-12 08:19:03+00:00 UTC
- store routes in arrays, not hashes, so that they can be defined in order
- add example2.pl

0.004     2014-10-12 06:00:25+00:00 UTC
- change 'delete' to 'del' since 'delete' is already a perl builtin function
- added support for multiple workers (default 5) and timeout (default 5 seconds)
- add "Server: limper/$Limper::VERSION" to response
- proper HEAD support
- dist.ini improvements and POD fix (thanks, garu!)
- remove dump_request(), and hence dep on Data::Dumper, so now this fatpacks nicely
- replaced Test::HTTP with the thing Limper was originally written to test

0.003     2014-09-21 03:37:31+00:00 UTC
- skip 002_limper.t if Test::HTTP not installed
- POD updates

0.002     2014-09-17 06:44:00+00:00 UTC
- silly error in POD

0.001     2014-09-17 06:36:36+00:00 UTC


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "extremely lightweight but not very powerful web application framework",
  "AUTHOR" => "Ashley Willis <ashley+perl\@gitable.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Limper",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Limper",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "IO::Socket" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.012",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "IO::Socket" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Limper/Engine.pm
package Limper::Engine;
$Limper::Engine::VERSION = '0.012';
use 5.10.0;
use strict;
use warnings;

1;

__END__

=head1 NAME

Limper::Engine - placeholder for the Limper::Engine namespace

=head1 VERSION

version 0.012

=head1 SYNOPSIS

Do not use this module directly. It currently does absolutely nothing, and likely never will.

=head1 DESCRIPTION

This namespace is specifically for plugins that connect L<Limper> to a
particular web server or protocol (like L<Limper::Engine::PSGI>), using
hooks L<Limper/request_handler> and L<Limper/response_handler>.

See L<Limper::Extending> for how to write L<Limper> plugins.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Ashley Willis E<lt>ashley+perl@gitable.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.


lib/Limper/Passphrase.pm
package Limper::Passphrase;
$Limper::Passphrase::VERSION = '0.012';
use 5.10.0;
use strict;
use warnings;

1;

__END__

=head1 NAME

Limper::Passphrase - generate and use passphrases with Limper

=head1 VERSION

version 0.012

=head1 SYNOPSIS

  use Authen::Passphrase;
  use Authen::Passphrase::BlowfishCrypt;
  #use Limper::Passphrase;
  use Limper;               # this must come after all extensions

  post '/password' => {
    # get $passphrase and $user somehow
    my $ppr = Authen::Passphrase::BlowfishCrypt->new(cost => 8, salt_random => 1, passphrase => $passphrase);
    # store $ppr->as_rfc2307 for user
  };

  post '/login' => {
    # get $passphrase and $user somehow
    if (Authen::Passphrase->from_rfc2307($user->{passphrase})->match($passphrase)) {
      # set some cookies or something
      'Logged in';
    } else {
      status 401;
      'Unauthorized';
    }
  };

  limp;

=head1 DESCRIPTION

B<Limper::Passphrase> shows you how to deal with passphrases with L<Limper>,
and nothing more.  You might as well use L<M>, but this has documentation.

This also serves to show that you B<don't need to make everything a plugin
for Limper>.  If it's not messing with the request, response, or hooks, it
should not be a plugin.

=head1 EXPORTS

Nothing additional is exported.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Ashley Willis E<lt>ashley+perl@gitable.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=head1 SEE ALSO

L<Limper>

L<Limper::Extending>

L<M>

=cut


lib/Limper.pm
package Limper;
$Limper::VERSION = '0.012';
use 5.10.0;
use strict;
use warnings;

use IO::Socket;

use Exporter qw/import/;
our @EXPORT = qw/get post put del trace options patch any status headers request response config hook limp/;
our @EXPORT_OK = qw/info warning rfc1123date/;

# data stored here
my $request = {};
my $response = {};
my $config = {};
my $hook = {};
my $conn;

# route subs
my $route = {};
sub get     { push @{$route->{GET}},     @_; @_ }
sub post    { push @{$route->{POST}},    @_; @_ }
sub put     { push @{$route->{PUT}},     @_; @_ }
sub del     { push @{$route->{DELETE}},  @_; @_ }
sub trace   { push @{$route->{TRACE}},   @_; @_ }
sub options { push @{$route->{OPTIONS}}, @_; @_ }
sub patch   { push @{$route->{PATCH}},   @_; @_ }
sub any     { push @{$route->{$_}},      @_ for keys %$route }

# for send_response()
my $reasons = {
    100 => 'Continue',
    101 => 'Switching Protocols',
    200 => 'OK',
    201 => 'Created',
    202 => 'Accepted',
    203 => 'Non-Authoritative Information',
    204 => 'No Content',
    205 => 'Reset Content',
    206 => 'Partial Content',
    300 => 'Multiple Choices',
    301 => 'Moved Permanently',
    302 => 'Found',
    303 => 'See Other',
    304 => 'Not Modified',
    305 => 'Use Proxy',
    307 => 'Temporary Redirect',
    400 => 'Bad Request',
    401 => 'Unauthorized',
    402 => 'Payment Required',
    403 => 'Forbidden',
    404 => 'Not Found',
    405 => 'Method Not Allowed',
    406 => 'Not Acceptable',
    407 => 'Proxy Authentication Required',
    408 => 'Request Time-out',
    409 => 'Conflict',
    410 => 'Gone',
    411 => 'Length Required',
    412 => 'Precondition Failed',
    413 => 'Request Entity Too Large',
    414 => 'Request-URI Too Large',
    415 => 'Unsupported Media Type',
    416 => 'Requested range not satisfiable',
    417 => 'Expectation Failed',
    500 => 'Internal Server Error',
    501 => 'Not Implemented',
    502 => 'Bad Gateway',
    503 => 'Service Unavailable',
    504 => 'Gateway Time-out',
    505 => 'HTTP Version not supported',
};

# for get_request()
my $method_rx = qr/(?: OPTIONS | GET | HEAD | POST | PUT | DELETE | TRACE | CONNECT )/x;
my $version_rx = qr{HTTP/\d+\.\d+};
my $uri_rx = qr/[^ ]+/;

# Returns current time or passed timestamp as an HTTP 1.1 date
my @months = qw/Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec/;
my @days = qw/Sun Mon Tue Wed Thu Fri Sat/;
sub rfc1123date {
    my ($sec, $min, $hour, $mday, $mon, $year, $wday) = @_ ? gmtime $_[0] : gmtime;
    sprintf '%s, %02d %s %4d %02d:%02d:%02d GMT', $days[$wday], $mday, $months[$mon], $year + 1900, $hour, $min, $sec;
}

# Formats date like "2014-08-17 00:12:41" in local time.
sub date {
    my ($sec, $min, $hour, $mday, $mon, $year) = localtime;
    sprintf '%04d-%02d-%02d %02d:%02d:%02d', $year + 1900, $mon, $mday, $hour, $min, $sec;
}

# Trivially log to STDOUT or STDERR
sub info    { say  date, ' ', @_ }
sub warning { warn date, ' ', @_ }

sub timeout {
    eval {
        local $SIG{ALRM} = sub { die "alarm\n" };
        alarm($config->{timeout} // 5);
        $_ = $_[0]->();
        alarm 0;
    };
    $@ ? ($conn->close and undef) : $_;
}

sub bad_request {
    warning "[$request->{remote_host}] bad request: $_[0]";
    $response = { status => 400, body => 'Bad Request' };
    send_response($request->{method} // '' eq 'HEAD', 'close');
}

# Returns a processed request as a hash, or sends a 400 and closes if invalid.
sub get_request {
    $request = { headers => {}, remote_host => $conn->peerhost // 'localhost' };
    $response = { headers => {} };
    my ($request_line, $headers_done, $chunked);
    while (1) {
        defined(my $line = timeout(sub { $conn->getline })) or last;
        if (!defined $request_line) {
            next if $line eq "\r\n";
            ($request->{method}, $request->{uri}, $request->{version}) = $line =~ /^($method_rx) ($uri_rx) ($version_rx)\r\n/;
            return bad_request $line unless defined $request->{method};
            $request_line = 1;
        } elsif (!defined $headers_done) {
            if ($line =~ /^\r\n/) {
                $headers_done = 1;
            } else {
                my ($name, $value) = split /:[ \t]*/, $line, 2;
                if ($name =~ /\r\n/) {
                    return bad_request $line;
                }
                $value =~ s/\r\n//;
                $value = $1 if lc $name eq 'host' and $request->{version} eq 'HTTP/1.1' and $request->{uri} =~ s{^https?://(.+?)/}{/};
                if (exists $request->{headers}{lc $name}) {
                    if (ref $request->{headers}{lc $name}) {
                        push @{$request->{headers}{lc $name}}, $value;
                    } else {
                        $request->{headers}{lc $name} = [$request->{headers}{lc $name}, $value];
                    }
                } else {
                    $request->{headers}{lc $name} = $value;
                }
            }
        }
        if (defined $headers_done) {
            return if defined $chunked;
            info "[$request->{remote_host}] $request->{method} $request->{uri} $request->{version} [", $request->{headers}{'user-agent'} // '', ']';
            return bad_request 'Host header missing' if $request->{version} eq 'HTTP/1.1' and (!exists $request->{headers}{host} or ref $request->{headers}{host});
            for (keys %{$request->{headers}}) {
                if ($_ eq 'expect' and lc $request->{headers}{$_} eq '100-continue' and $request->{version} eq 'HTTP/1.1') {
                    $conn->print("HTTP/1.1 100 Continue\r\n\r\n");	# this does not check if route is valid. just here to comply.
                } elsif ($_ eq 'content-length') {
                    timeout(sub { $conn->read($request->{body}, $request->{headers}{$_}) });
                    last;
                } elsif ($_ eq 'transfer-encoding' and lc $request->{headers}{$_} eq 'chunked') {
                    my $length = my $offset = $chunked = 0;
                    do {
                        $_ = timeout(sub { $conn->getline });
                        $length = hex((/^([A-Fa-f0-9]+)(?:;.*)?\r\n/)[0]);
                        timeout(sub { $conn->read($request->{body}, $length + 2, $offset) }) if $length;
                        $offset += $length;
                    } while $length;
                    $request->{body} =~ s/\r\n$//;
                    undef $headers_done; # to get optional footers, and another blank line
                }
            }
            last if defined $headers_done;
        }
    }
}

# Finds and calls the appropriate route sub, or sends a 404 response.
sub handle_request {
    my $head = 1;
    (defined $request->{method} and $request->{method} eq 'HEAD') ? ($request->{method} = 'GET') : ($head = 0);
    if (defined $request->{method} and exists $route->{$request->{method}}) {
        for (my $i = 0; $i < @{$route->{$request->{method}}}; $i += 2) {
            if ($route->{$request->{method}}[$i] eq $request->{uri} ||
                        ref $route->{$request->{method}}[$i] eq 'Regexp' and $request->{uri} =~ $route->{$request->{method}}[$i]) {
                $response->{body} = & { $route->{$request->{method}}[$i+1] };
                return send_response($head);
            }
        }
    }
    $response->{body} = 'This is the void';
    $response->{status} = 404;
    send_response($head);
}

# Sends a response to client. Default status is 200.
sub send_response {
    my ($head, $connection) = @_;
    $connection //= (($request->{version} // '') eq 'HTTP/1.1')
            ? lc($request->{headers}{connection} // '')
            : lc($request->{headers}{connection} // 'close') eq 'keep-alive' ? 'keep-alive' : 'close';
    $response->{status} //= 200;
    $response->{headers}{Date} = rfc1123date();
    if (defined $response->{body} and !ref $response->{body}) {
        $response->{headers}{'Content-Length'} //= length $response->{body};
        $response->{headers}{'Content-Type'} //= 'text/plain';
    }
    delete $response->{body} if $head // 0;
    $response->{headers}{Connection} = $connection if $connection eq 'close' or ($connection eq 'keep-alive' and $request->{version} ne 'HTTP/1.1');
    $response->{headers}{Server} = 'limper/' . ($Limper::VERSION // 'pre-release');
    $_->($request, $response) for @{$hook->{after}};
    return $hook->{response_handler}[0]->() if exists $hook->{response_handler};
    {
        local $\ = "\r\n";
        $conn->print(join ' ', $request->{version} // 'HTTP/1.1', $response->{status}, $response->{reason} // $reasons->{$response->{status}});
        return unless $conn->connected;
        my @headers = headers();
        $conn->print( join(': ', splice(@headers, 0, 2)) ) while @headers;
        $conn->print();
    }
    $conn->print($response->{body} // '') if defined $response->{body};
    $conn->close if $connection eq 'close';
}

sub status {
    if (defined wantarray) {
        wantarray ? ($response->{status}, $response->{reason}) : $response->{status};
    } else {
        $response->{status} = shift;
        $response->{reason} = shift if @_;
    }
}

sub headers {
    if (!defined wantarray) {
        $response->{headers}{+pop} = pop while @_;
    } else {
        my @headers;
        for my $key (keys %{ $response->{headers} }) {
            if (ref $response->{headers}{$key}) {
                push @headers, $key, $_ for @{$response->{headers}{$key}};
            } else {
                push @headers, $key, $response->{headers}{$key};
            }
        }
        @headers;
    }
}

sub request { $request }

sub response { $response }

sub config { $config }

sub hook { push @{$hook->{$_[0]}}, $_[1] }

sub limp {
    $config = shift @_ if ref $_[0] eq 'HASH';
    return $hook->{request_handler}[0] if exists $hook->{request_handler};
    my $sock = IO::Socket::INET->new(Listen => SOMAXCONN, ReuseAddr => 1, LocalAddr => 'localhost', LocalPort => 8080, Proto => 'tcp', @_)
            or die "cannot bind to port: $!";

    info 'limper started';

    for (1 .. $config->{workers} // 5) {
        defined(my $pid = fork) or die "fork failed: $!";
        while (!$pid) {
            if ($conn = $sock->accept()) {
                do {
                    eval {
                        get_request;
                        handle_request if $conn->connected;
                    };
                    if ($@) {
                        $response = { status => 500, body => $config->{debug} // 0 ? $@ : 'Internal Server Error' };
                        send_response 0, 'close';
                        warning $@;
                    }
                } while ($conn->connected);
            }
        }
    }
    1 while (wait != -1);

    my $shutdown = $sock->shutdown(2);
    my $closed = $sock->close();
    info 'shutdown ', $shutdown ? 'successful' : 'unsuccessful';
    info 'closed ', $closed ? 'successful' : 'unsuccessful';
}

1;

__END__

=for Pod::Coverage bad_request date get_request handle_request send_response timeout

=head1 NAME

Limper - extremely lightweight but not very powerful web application framework

=head1 VERSION

version 0.012

=head1 SYNOPSIS

  use Limper;

  my $generic = sub { 'yay' };

  get '/' => $generic;
  post '/' => $generic;

  post qr{^/foo/} => sub {
      status 202, 'whatevs';
      headers Foo => 'bar', Fizz => 'buzz';
      'you posted something: ' . request->{body};
  };

  limp;

=head1 DESCRIPTION

B<Limper> was originally designed to primarily be a simple HTTP/1.1 test
server in perl, but I realized it can be much more than that while still
remaining simple.

B<Limper> has a simple syntax like L<Dancer> yet no dependencies at all,
unlike the dozens that L<Dancer> pulls in.

B<Limper> is modular - support for serving files, easily returning JSON, or
using PSGI can be included if and only if needed (and these features already
exist on CPAN).

B<Limper> is fast - about 2-3 times faster than Dancer.

B<Limper> also fatpacks beautifully (at least on 5.10.1):

  fatpack pack example.pl > example-packed.pl

Do not taunt B<Limper>.

=head1 EXPORTS

The following are all exported by default:

  get post put del trace options patch any
  status headers request response config hook limp

Also exportable:

  info warning rfc1123date

=head1 FUNCTIONS

=head2 get

=head2 post

=head2 put

=head2 del

=head2 trace

=head2 options

=head2 patch

Defines a route handler for METHOD to the given path:

  get '/' => sub { 'Hello world!' };

These can be chained together like so:

  get post del '/' => sub { 'Hello world!' };

Note that a route to match B<HEAD> requests is automatically created as well for B<get>.

=head2 any

Defines a route handler for B<all> METHODs to the given path:

  any '/' => sub { 'Hello world!' };

=head2 status

Get or set the response status, and optionally reason.

  status 404;
  status 401, 'Nope';
  my $status = status;
  my ($status, $reason) = status;

=head2 headers

Get or set the response headers.

  headers Foo => 'bar', Fizz => 'buzz';
  headers Foo => ['this', 'that'];                # change Foo to two values
  headers Oops => 'inserted', Oops => 'ignored';  # don't do this
  my @headers = headers;

Note: Changed in 0.012. The headers are now stored in a hashref, where the
value is either a string or array of strings.  Calling B<headers> in list
mode returns a flattened list.  If you want the hashref of headers, use B<
response->{headers} >>.  Setting header pairs no longer overwrites all
previously defined headers.

=head2 request

Returns a B<HASH> of the request. Request keys are: B<method>, B<uri>, and
B<version>.  It may also contain B<headers> which is a B<HASH> and B<body>.

There is no decoding of the body content nor URL paramters.

=head2 response

Returns response B<HASH>. Keys are B<status>, B<reason>, B<headers> (a
B<HASH> of key/value pairs), and B<body>.

=head2 config

Returns config B<HASH>. See B<limp> below for known config settings.

=head2 hook

Adds a hook at some position.

Three hooks are currently defined: B<after>, B<request_handler>, and B<response_handler>.

=head3 after

Runs after all other processing, just before response is sent.

  hook after => sub {
    my ($request, $response) = @_;
    # modify response as needed
  };

=head3 request_handler

Runs when B<limp> is called, after only setting passed config settings, and returns
the result instead of starting up the built-in web server.  A simplified
example for PSGI (including the B<response_handler> below) is:

  hook request_handler => sub {
    get_psgi @_;
    handle_request;
  };

=head3 response_handler

Runs right after the B<after> hook, and returns the result instead of using
the built-in web server for sending the response. For PSGI, this is:

  hook response_handler => sub {
    [ response->{status}, [headers], ref response->{body} ? response->{body} : [response->{body}] ];
  };

=head2 limp

Starts the server. You can pass it the same options as L<IO::Socket::INET>
takes.  The default options are:

  Listen => SOMAXCONN, ReuseAddr => 1, LocalAddr => 'localhost', LocalPort => 8080, Proto => 'tcp'

In addition, the first argument can be a B<HASH> to pass config settings:

  limp({debug => 1, timeout => 60, workers => 10}, LocalAddr => '0.0.0.0', LocalPort => 3001);

Default debug is B<0>, default timeout is B<5> (seconds), and default
workers is B<10>.  A timeout of B<0> means never timeout.

This keyword should be called at the very end of the script, once all routes
are defined.  At this point, Limper takes over control.

=head1 ADDITIONAL FUNCTIONS

=head2 info

=head2 warning

Log given list to B<STDOUT> or B<STDERR>. Prepends the current local time in
format "YYYY-MM-DD HH:MM:SS".

=head2 rfc1123date

Returns the current time or passed timestamp as an HTTP 1.1 date (RFC 1123).

=head1 EVEN MORE

For additional (discouraged) functions to aid in transitioning to Limper, see L<Limper::Sugar>.

For sending files and easily sending JSON, see L<Limper::SendFile> and L<Limper::SendJSON>.

For differences between Limper and Dancer, see L<Limper::Differences>.

For extending Limper, see L<Limper::Extending>.

=head1 NOTICE

This framework is still under development. Things B<may> change without
warning.  Version 0.012 has such changes, but I hope I have what is in this
version stable.

=head1 BREAKING CHANGES IN 0.012

B<options> is now B<config>, and there is a new function B<options> for the
HTTP method.

B<note> has been changed to B<info>.

B<headers> now will update just the fields given, and not replace all the
headers.  The headers are now stored as a B<HASH> instead of an B<ARRAY>.
Hence, C<< response->{headers} >> cannot be directly passed to PSGI.  Instead
C<< [headers] >> meets this need.

C<< request->{header} >> is now what C<< request->{hheader} >> was - no more
ARRAY form.

=head1 CONTRIBUTING

=head2 Patches and Bug Fixes

Preferably, clone the repo (uses L<Dist::Zilla>) and create one or more
patch files with:

  git format patch <latest commit>

Email me the patch, or otherwise let me know how to find it.

Or if it's a simple patch and you don't want to mess with L<Dist::Zilla>,
patch the latest release and send me a patch file.

=head2 Module Namespaces

See L<Limper::Extending/NAMESPACES>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Ashley Willis E<lt>ashley+perl@gitable.orgE<gt>

B<rabcyr> on irc and L<twitter|https://twitter.com/rabcyr>.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=head1 SEE ALSO

L<Limper::Differences>

L<Limper::Extending>

L<Limper::Engine::PSGI>

L<Limper::SendFile>

L<Limper::SendJSON>

L<Limper::Sugar>

L<App::FatPacker>

L<IO::Socket::INET>

L<Web::Simple>
